SPI Chip Select tidy-up

pelwell opened the pull request:
These three commits, together with the previous "spi: Force CS_HIGH if GPIO descriptors are used", tidy up the handling of SPI Chip Select signals, particularly for controllers that use GPIO descriptors. The problems observed since the end of 2019 stem from the setting too early of SPI_CS_HIGH flag for all CS GPIOs managed by descriptors. The earlier commit moves the flag setting later, and two of these commits remove premature settings.

The third commit avoids a glitch on active-low CS lines on GPIOs that naturally pull high by declaring the pins to pinctrl as inputs.

pelwell committed:
spi: use_gpio_descriptor fixup moved to spi_setup

Commits [1] and [2] including code that forces SPI_CS_HIGH for SPI
controllers that use GPIO descriptors, the SPI_CS_HIGH flag being
there to avoid a double-negation (since SPI CS is usually active-low).
The motivation for pushing the knowledge of the required polarity into
the GPIO descriptor allows the switch to an output to request the
correct inactive level, avoiding a needless glitch.

The problem with setting the flag early as [1] does is that it appears
in the mode field that is passed to client drivers during their probing,
when they may want to choose SPI_POL, SPI_PHA and (just possibly)
SPI_CS_HIGH. Since SPI_CS_HIGH is the exception, most drivers won't
set it and the anti-negation negation is lost. [2] acknowledges that
problem and patches things up for the case of users of spidev, but
omits regular kernel-mode drivers.

Downstream commit [3] moves the forcing of SPI_CS_HIGH to spi_setup,
after the driver probing. Since this code is called before any CS
manipulation it is early enough to be effective, but late enough that
clients have already had their chance to change the mode field.

This is a partial reversion of [1], and is accompanied by a complete
reversion of [2], neither of which is needed any longer.

[1] f3186dd87669 ("spi: Optionally use GPIO descriptors for CS GPIOs")
[2] 83b2a8fe43bd ("spi: spidev: Fix CS polarity if GPIO descriptors are used")
[3] <varies> ("spi: Force CS_HIGH if GPIO descriptors are used")

Signed-off-by: Phil Elwell <phil@raspberrypi.com>

pelwell committed:
Revert "spi: spidev: Fix CS polarity if GPIO descriptors are used"

This reverts commit 83b2a8fe43bda0c11981ad6afa5dd0104d78be28.

pelwell commented:
@6by9 Judging by the waveforms from the logic analyser I can't imagine the TV HAT won't work after this is merged, but some confirmation would be nice.

HiassofT commented:
I understand the (potential) problem with SPI chip select glitching but declaring them as inputs in the device tree (which describes hardware) and then hoping that the drivers actually configure them as outputs (opposed to DT config) looks very wrong to me.

SPI devices should return to a reset I/O state as soon as CS is de-asserted (i.e. when spi controller initializes the gpiod lines) so this rather crude workaround shouldn't be needed in general.

If initial pin output state is needed in pinctrl then it should be added there.

pelwell commented:
These are GPIO CS declarations - it's the core SPI code and GPIO descriptor setup code that sets them to be outputs with the correct initial level.

HiassofT commented:
yes, I know, but pinctrl settings are pulled in via pinmux configs and setting pinctrl to input doesn't look correct to me.

might be a good idea to run this through the upstream gpio / DT folks, I'm pretty sure people on other platforms may have run into that issue too (I'm not sure what the best or a correct solution would be to avoid these glitches)

pelwell commented:
It's probably less contentious just to delete the pinctrl declarations - they serve little purpose in this case.

6by9 commented:
I'll try to test this on the TV HAT this morning. Nag me if I don't get back to you.

6by9 commented:
TV HAT identified and has found at least 498 and 522 MHz so far. Looks good.

pelwell commented:
Thanks.

6by9 commented:
61 channels found on my dodgy aerial feed. I think that was the same as before.
```
service_id 8258;(null):498000:B8C23D0G32M64T8Y0:T:27500:1701=2:1702=eng@4,1703=eng:0;1731:0:8258:9018:8194:0
service_id 8294;(null):498000:B8C23D0G32M64T8Y0:T:27500:2301=2:2302=eng@4,2303=eng:0;2331:0:8294:9018:8194:0
service_id 8325;(null):498000:B8C23D0G32M64T8Y0:T:27500:1801=2:1802=eng@4,1803=eng:0;1831:0:8325:9018:8194:0
service_id 8330;(null):498000:B8C23D0G32M64T8Y0:T:27500:2001=2:2002=eng@4,2003=eng:0;2031:0:8330:9018:8194:0
service_id 8357;(null):498000:B8C23D0G32M64T8Y0:T:27500:1901=2:1902=eng@4,1903=eng:0;1931:0:8357:9018:8194:0
service_id 8384;(null):498000:B8C23D0G32M64T8Y0:T:27500:1101=2:1102=eng@4,1103=eng:0;1131:0:8384:9018:8194:0
service_id 8385;(null):498000:B8C23D0G32M64T8Y0:T:27500:1501=2:1502=eng@4,1503=eng:0;1531:0:8385:9018:8194:0
service_id 8442;(null):498000:B8C23D0G32M64T8Y0:T:27500:1401=2:1402=eng@4,1403=eng:0;1431:0:8442:9018:8194:0
service_id 8448;(null):498000:B8C23D0G32M64T8Y0:T:27500:1301=2:1302=eng@4,1303=eng:0;1331:0:8448:9018:8194:0
service_id 8452;(null):498000:B8C23D0G32M64T8Y0:T:27500:1201=2:1202=eng@4,1203=eng:0;1231:0:8452:9018:8194:0
service_id 8458;(null):498000:B8C23D0G32M64T8Y0:T:27500:2101=2:2102=eng@4,2103=eng:0;2131:0:8458:9018:8194:0
service_id 8500;(null):498000:B8C23D0G32M64T8Y0:T:27500:1601=2:1602=eng@4,1603=eng:0;1631:0:8500:9018:8194:0
5STAR;(null):570000:B8C34D0G32M64T8Y0:T:27500:6673=2:6674=eng@3,6675=eng:0;6678:0:12928:9018:12294:0
5 USA;(null):570000:B8C34D0G32M64T8Y0:T:27500:6689=2:6690=eng@3,6691=eng:0;6694:0:12992:9018:12294:0
Create & Craft;(null):570000:B8C34D0G32M64T8Y0:T:27500:6161=2:6162=eng@3:0:0:13000:9018:12294:0
Channel 5+1;(null):570000:B8C34D0G32M64T8Y0:T:27500:6865=2:6866=eng@3,6867=eng:0;6870:0:13024:9018:12294:0
QVC;(null):570000:B8C34D0G32M64T8Y0:T:27500:6049=2:6050=eng@3:0:0:13120:9018:12294:0
Blaze;(null):570000:B8C34D0G32M64T8Y0:T:27500:6065=2:6066=eng@3:0:0:14384:9018:12294:0
Hochanda;(null):570000:B8C34D0G32M64T8Y0:T:27500:6705=2:6706=eng@3:0:0:14412:9018:12294:0
CBS Reality;(null):570000:B8C34D0G32M64T8Y0:T:27500:6289=2:6290=eng@3,6291=eng:0;6294:0:14448:9018:12294:0
Horror Channel;(null):570000:B8C34D0G32M64T8Y0:T:27500:6129=2:6130=eng@3,6131=eng:0;6134:0:14480:9018:12294:0
QUEST;(null):570000:B8C34D0G32M64T8Y0:T:27500:6929=2:6930=eng@3,6931=eng:0;6934:0:14498:9018:12294:0
Absolute Radio;(null):570000:B8C34D0G32M64T8Y0:T:27500:0:6082=eng@3:0:0:14688:9018:12294:0
Heart;(null):570000:B8C34D0G32M64T8Y0:T:27500:0:6098=eng@3:0:0:14720:9018:12294:0
Capital;(null):570000:B8C34D0G32M64T8Y0:T:27500:0:6114=eng@3:0:0:14752:9018:12294:0
Sony Movies Action;(null):570000:B8C34D0G32M64T8Y0:T:27500:6913=2:6914=eng@3,6915=eng:0;6918:0:15576:9018:12294:0
TJC;(null):570000:B8C34D0G32M64T8Y0:T:27500:6145=2:6146=eng@3:0:0:15592:9018:12294:0
ITVBe;(null):570000:B8C34D0G32M64T8Y0:T:27500:6497=2:6498=eng@3,6499=eng:0;6502:0:15856:9018:12294:0
ITV2 +1;(null):570000:B8C34D0G32M64T8Y0:T:27500:6529=2:6530=eng@3,6531=eng:0;6534:0:15952:9018:12294:0
CITV;(null):570000:B8C34D0G32M64T8Y0:T:27500:6833=2:6834=eng@3,6835=eng:0;6838:0:16032:9018:12294:0
5SELECT;(null):570000:B8C34D0G32M64T8Y0:T:27500:6657=2:6658=eng@3,6659=eng:0;6662:0:16080:9018:12294:0
Paramount Network;(null):570000:B8C34D0G32M64T8Y0:T:27500:6593=2:6594=eng@3,6595=eng:0;6598:0:16112:9018:12294:0
Drama;(null):570000:B8C34D0G32M64T8Y0:T:27500:6609=2:6610=eng@3,6611=eng:0;6614:0:16208:9018:12294:0
That's TV;(null):586000:B8C34D0G32M2T8Y0:T:27500:3201=2:3202=eng@3:0:0:32858:9018:32794:0
Sony Channel;(null):586000:B8C34D0G32M2T8Y0:T:27500:1401=2:1402=eng@3:0:0:33984:9018:32794:0
Sony Movies Classic;(null):586000:B8C34D0G32M2T8Y0:T:27500:1101=2:1102=eng@3:0:0:34048:9018:32794:0
Tiny Pop;(null):586000:B8C34D0G32M2T8Y0:T:27500:4901=2:4902=eng@3:0:0:33472:9018:32794:0
BBC ONE East W;(null):522000:B8C23D0G32M64T8Y0:T:27500:101=2:102=eng@3,106=eng:0;105:0:4173:9018:4173:0
BBC TWO;(null):522000:B8C23D0G32M64T8Y0:T:27500:201=2:202=eng@3,206=eng:0;205:0:4287:9018:4173:0
BBC NEWS;(null):522000:B8C23D0G32M64T8Y0:T:27500:501=2:502=eng@3:0;505:0:4352:9018:4173:0
CBBC;(null):522000:B8C23D0G32M64T8Y0:T:27500:301=2:302=eng@3,306=eng:0;305:0:4608:9018:4173:0
CBeebies;(null):522000:B8C23D0G32M64T8Y0:T:27500:401=2:402=eng@3,406=eng:0;405:0:4672:9018:4173:0
BBC Parliament;(null):522000:B8C23D0G32M64T8Y0:T:27500:601=2:602=eng@3:0;605:0:4736:9018:4173:0
BBC R5L;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1402=eng@3:0:0:5632:9018:4173:0
BBC R5SX;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1502=eng@3:0:0:5696:9018:4173:0
BBC 6 Music;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1602=eng@3:0:0:5760:9018:4173:0
BBC Radio 4 Ex;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1702=eng@3:0:0:5824:9018:4173:0
BBC R1X;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1802=eng@3:0:0:5888:9018:4173:0
BBC Asian Net.;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1902=eng@3:0:0:5952:9018:4173:0
BBC World Sv.;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:2002=eng@3:0:0:6016:9018:4173:0
BBC Radio London;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:3802=eng@3:0:0:6148:9018:4173:0
BBC Norfolk;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:3202=eng@3:0:0:6151:9018:4173:0
BBC Three Counties;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:3602=eng@3:0:0:6157:9018:4173:0
BBC Northampton;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:3702=eng@3:0:0:6221:9018:4173:0
BBC Cambridge;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:3502=eng@3:0:0:6285:9018:4173:0
BBC Radio 1;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1002=eng@3:0:0:6720:9018:4173:0
BBC Radio 2;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1102=eng@3:0:0:6784:9018:4173:0
BBC Radio 3;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1202=eng@3:0:0:6848:9018:4173:0
BBC Radio 4;(null):522000:B8C23D0G32M64T8Y0:T:27500:0:1302=eng@3:0:0:6912:9018:4173:0
BBC RB 1;(null):522000:B8C23D0G32M64T8Y0:T:27500:901=2:951=eng@3,952=und:0:0:7168:9018:4173:0
Done, scan time: 11:18.692
```

pelwell committed:
ARM: dts: bcm27xx: Remove SPI CS pins from pinctrl

SPI Chip Selects are output signals, but declaring them to pinctrl as
outputs forces them to low even though they are almost certainly going
be active-low signals. This causes a brief glitch until the SPI
subsystem gets hold of them and switches them to outputs driven high.

Leaving the pins as inputs until the SPI driver claims them allows
their natural pulls to persist until an informed choice of level can
be made leaving, at worst, a single edge and, more likely, no change
at all.

The spi*_cs_pins nodes are left for backwards-compatibility with custom
overlays and HAT EEPROMs.

Signed-off-by: Phil Elwell <phil@raspberrypi.com>

pelwell commented:
Take two, this time removing the CS pins from pinctrl, but leaving placeholders in case there are overlays around that target them.

pelwell commented:
Are you happy for me to merge, @popcornmix?

popcornmix commented:
Yes
